### PR TITLE
working corner cases for 32 and 64 bit, direct and numeric conversions

### PR DESCRIPTION
**Pull request by desmonddak**
## Description & Motivation

32 and 64 bit corner cases test both the numeric and direct input conversion paths.

## Testing

Ran corner case values by converting into our FP representations, then back out and compare.

## Backwards-compatibility

> Is this a breaking change that will not be backwards-compatible? If yes, how so?

No

## Documentation

> Does the change require any updates to documentation? If so, where? Are they included?

Eventually we will update documentation on the FP value system and connection to FP logic types.

### CODE REVIEW

**Comment by desmonddak** [comment]
Using the numeric conversion path even though a direct copy path from host data is available, we test this path on standard widths to prepare its use for non-standard floating point widths.  There may be semi-direct paths available for shorter widths (say by copying a truncation of the mantissa and mapping the exponent to a different bias), the numeric conversion path will serve as a baseline.

File: lib/src/adder.dart [hunk 1]
```diff
@@ -66,3 +74,48 @@ class FullAdder extends Module {
...
+      : super(name: 'Ones Complement Adder') {
+    final aOnesComplement = Logic(width: a.width);
+    final bOnesComplement = Logic(width: b.width);
+    final sign = addOutput('sign');
+
+    aOnesComplement <= mux(aSign, ~a, a);
+    bOnesComplement <= mux(bSign, ~b, b);
+
+    final adder = adderGen(aOnesComplement, bOnesComplement);
+    // print('\tA  ${aOnesComplement.value.toString(includeWidth: false)}');
```
Comment (mkorbel1) [inline lib/src/adder.dart]: could get rid of commented out/print stuff

File: lib/src/floating_point_value.dart [hunk 2]
```diff
@@ -0,0 +1,482 @@
...
+      mantissa: mantissa);
+}
+
+/// A representation of a single precision floating point value
+class FloatingPoint32Value extends FloatingPointValue {
+  static const int _exponentWidth = 8;
+  static const int _mantissaWidth = 23;
+
+  /// return the exponent width
+  static int get exponentWidth => _exponentWidth;
```
Comment (mkorbel1) [inline lib/src/floating_point_value.dart]: you could just make the `_exponentWidth` and `_mantissaWidth` public directly instead of providing getters since they are `const` and can't be changed anyways.

File: lib/src/floating_point.dart [hunk 3]
```diff
@@ -246,24 +45,122 @@ class FloatingPoint extends LogicStructure {
...
+        ..gets(output('out'));
+
+  /// The result of [FloatingPoint] addition
+  // FloatingPoint get out => output('out');
+  @protected
+  late final FloatingPoint _out =
+      FloatingPoint(exponentWidth: exponentWidth, mantissaWidth: mantissaWidth);
+
+  /// Swapping two FloatingPoint structures based on a conditional
+  static (FloatingPoint, FloatingPoint) swap(
```
Comment (mkorbel1) [inline lib/src/floating_point.dart]: this doesn't need to be public (rename to `_swap`)

File: lib/src/ripple_carry_adder.dart [hunk 4]
```diff
@@ -32,5 +38,35 @@ class RippleCarryAdder extends Adder {
     sumList.add(carry);
 
     sum <= sumList.rswizzle();
+    carryOut <= sum[sum.width - 1];
+    out <= sum.slice(sum.width - 2, 0);
+  }
+}
+
+/// A ripple carry adder with a carry in
+class RippleCarryAdder2 extends Adder {
```
Comment (mkorbel1) [inline lib/src/ripple_carry_adder.dart]: why does this need to be a separate class? why not make `carry` an optional named argument?

File: test/adder_test.dart [hunk 5]
```diff
@@ -0,0 +1,171 @@
+// Copyright (C) 2023 Intel Corporation
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// adder_test.dart
+// Tests for the Adder interface.
+//
+// 2024 April 4
+// Author: Desmond Kirkpatrick <desmond.a.kirkpatrick@intel.com>
+
+// ignore_for_file: avoid_print, unnecessary_parenthesis
```
Comment (mkorbel1) [inline test/adder_test.dart]: avoid these lint waivers

**Comment by desmonddak** [comment]
I added floating point to the web generator and tested it locally.  I'm working next on documentation and a generated schematic.

File: lib/src/adder.dart [hunk 6]
```diff
@@ -21,6 +21,12 @@ abstract class Adder extends Module {
   @protected
   late final Logic b;
 
+  /// The addition results [out].
+  Logic get out => output('out');
```
Comment (mkorbel1) [inline lib/src/adder.dart:25]: Perhaps it's a good idea to add some un-implemented abstract functions to ensure that all implementations provide out, sum, and carryOut?

e.g.
https://github.com/intel/rohd-hcl/blob/main/lib/src/error_checking/error_checker.dart

File: lib/src/adder.dart [hunk 7]
```diff
@@ -66,3 +74,37 @@ class FullAdder extends Module {
...
   }
 }
+
+/// An Adder which performs one's complement arithmetic using an unsigned
+/// adder that is passed in using a functor
+///    -- Requires that if the larger magnitude number is negative it
+///       must be the first 'a' argument
+///       We cannot enforce because this may be a smaller mantissa in
+///       a larger magnitude negative number (no asserts please)
+class OnesComplementAdder extends Adder {
```
Comment (mkorbel1) [inline lib/src/adder.dart]: This could be in a separate file

File: lib/src/floating_point_value.dart
Comment (mkorbel1) [inline lib/src/floating_point_value.dart:1]: maybe time to add a new "arithmetic" folder, and a "floating_point" folder under that, to keep things more organized than just top-level src with lots of files?

File: lib/src/adder.dart (hunk 6 above)
Comment (desmonddak) [reply lib/src/adder.dart:25]: I think I figured out this pattern

```dart
class RippleCarryAdder extends Adder {
  late final Logic _out;  // initialize in the constructor
  late final Logic _carry = Logic();

  /// Constructs an n-bit adder based on inputs List of inputs.
  RippleCarryAdder(super.a, super.b,
      {super.name = 'ripple_carry_adder', Logic? carry}) : _out = Logic(width: a.width) {
  ...
  }
  @override
  @protected
  Logic calculateOut() => _out;

  @override
  @protected
  Logic calculateCarry() => _carry;
}
```
When I run a test, it complains that _sum is uninitialized.

**Comment by desmonddak** [comment]
I updated according to your past three comments.
I still need to add some documentation, though.

**Review by mkorbel1 — APPROVED** [review]
Looks good!  I'll merge this into my branch and we can work on docs and other stuff next.  Seems almost ready to put into the main branch!
